File: surogate/server/routes/tasks.py
"""Local task API routes — spawn, list, cancel, logs."""

from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query, Request
from sqlalchemy.ext.asyncio import AsyncSession

import sqlalchemy as sa
from surogate.core.db.engine import get_session
from surogate.core.db.models.compute import LocalTask, LocalTaskStatus, LocalTaskType
from surogate.server.auth.authentication import get_current_subject
from surogate.server.models.tasks import (
    LocalTaskResponse,
    TaskLogsResponse,
    TaskSpawnRequest,
)

router = APIRouter()


def _to_response(t: LocalTask) -> LocalTaskResponse:
    return LocalTaskResponse(
        id=t.id,
        name=t.name,
        task_type=t.task_type.value,
        status=t.status.value,
        pid=t.pid,
        exit_code=t.exit_code,
        error_message=t.error_message,
        progress=t.progress,
        project_id=t.project_id,
        requested_by=t.requested_by_id,
        created_at=t.created_at,
        started_at=t.started_at,
        completed_at=t.completed_at,
    )


@router.get("", response_model=list[LocalTaskResponse])
async def list_tasks(
    current_subject: str = Depends(get_current_subject),
    session: AsyncSession = Depends(get_session),
    project_id: Optional[str] = Query(None),
    task_type: Optional[str] = Query(None),
    status: Optional[str] = Query(None),
    limit: int = Query(50, le=200),
):
    """List local tasks with optional filters."""
    stmt = sa.select(LocalTask).order_by(LocalTask.created_at.desc())
    if project_id:
        stmt = stmt.where(LocalTask.project_id == project_id)
    if task_type:
        stmt = stmt.where(LocalTask.task_type == LocalTaskType(task_type))
    if status:
        stmt = stmt.where(LocalTask.status == LocalTaskStatus(status))
    stmt = stmt.limit(limit)
    result = await session.execute(stmt)
    return [_to_response(t) for t in result.scalars().all()]


@router.post("", response_model=LocalTaskResponse)
async def spawn_task(
    req: TaskSpawnRequest,
    request: Request,
    current_subject: str = Depends(get_current_subject),
    session: AsyncSession = Depends(get_session),
):
    """Spawn a new local task as a subprocess."""
    manager = request.app.state.task_manager
    task = await manager.spawn(
        session,
        task_type=req.task_type,
        name=req.name,
        project_id=req.project_id,
        user_id=current_subject,
        params=req.params,
    )
    return _to_response(task)


@router.get("/{task_id}", response_model=LocalTaskResponse)
async def get_task(
    task_id: str,
    current_subject: str = Depends(get_current_subject),
    session: AsyncSession = Depends(get_session),
):
    """Get a single task's details."""
    result = await session.execute(
        sa.select(LocalTask).where(LocalTask.id == task_id)
    )
    task = result.scalar_one_or_none()
    if task is None:
        raise HTTPException(status_code=404, detail="Task not found")
    return _to_response(task)


@router.post("/{task_id}/cancel")
async def cancel_task(
    task_id: str,
    request: Request,
    current_subject: str = Depends(get_current_subject),
    session: AsyncSession = Depends(get_session),
):
    """Cancel a running task via SIGTERM."""
    manager = request.app.state.task_manager
    try:
        await manager.cancel(session, task_id)
    except ValueError as exc:
        raise HTTPException(status_code=404, detail=str(exc))
    return {"status": "cancelled"}


@router.delete("/{task_id}")
async def delete_task(
    task_id: str,
    request: Request,
    current_subject: str = Depends(get_current_subject),
    session: AsyncSession = Depends(get_session),
):
    """Cancel (if running) and delete a task record."""
    result = await session.execute(
        sa.select(LocalTask).where(LocalTask.id == task_id)
    )
    task = result.scalar_one_or_none()
    if task is None:
        raise HTTPException(status_code=404, detail="Task not found")
    if task.status in (LocalTaskStatus.pending, LocalTaskStatus.running):
        manager = request.app.state.task_manager
        try:
            await manager.cancel(session, task_id)
        except ValueError:
            pass
    await session.delete(task)
    await session.commit()
    return {"status": "deleted"}


@router.get("/{task_id}/logs", response_model=TaskLogsResponse)
async def get_task_logs(
    task_id: str,
    request: Request,
    current_subject: str = Depends(get_current_subject),
    tail: int = Query(100, le=1000),
):
    """Get the last N lines from a task's log file."""
    manager = request.app.state.task_manager
    lines = await manager.get_logs(task_id, tail=tail)
    return TaskLogsResponse(task_id=task_id, lines=lines)

File: frontend/src/api/tasks.ts
// Copyright (c) 2026, Invergent SA, developed by Flavius Burca
// SPDX-License-Identifier: AGPL-3.0-only
//
import { authFetch } from "@/api/auth";

export interface LocalTask {
  id: string;
  name: string;
  task_type: string;
  status: string;
  pid?: number | null;
  exit_code?: number | null;
  error_message?: string | null;
  progress?: string | null;
  project_id: string;
  requested_by: string;
  created_at: string;
  started_at?: string | null;
  completed_at?: string | null;
}

export interface TaskLogsResponse {
  task_id: string;
  lines: string[];
}

export async function spawnTask(params: {
  task_type: string;
  name: string;
  project_id: string;
  params: Record<string, string>;
}): Promise<LocalTask> {
  const response = await authFetch("/api/tasks", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(params),
  });
  if (!response.ok) {
    const err = (await response.json().catch(() => null)) as { detail?: string } | null;
    throw new Error(err?.detail ?? "Failed to spawn task");
  }
  return (await response.json()) as LocalTask;
}

export async function listTasks(params?: {
  project_id?: string;
  task_type?: string;
  status?: string;
}): Promise<LocalTask[]> {
  const qs = new URLSearchParams();
  if (params?.project_id) qs.append("project_id", params.project_id);
  if (params?.task_type) qs.append("task_type", params.task_type);
  if (params?.status) qs.append("status", params.status);
  const response = await authFetch(`/api/tasks?${qs}`);
  if (!response.ok) {
    const err = (await response.json().catch(() => null)) as { detail?: string } | null;
    throw new Error(err?.detail ?? "Failed to list tasks");
  }
  return (await response.json()) as LocalTask[];
}

export async function getTask(taskId: string): Promise<LocalTask> {
  const response = await authFetch(`/api/tasks/${taskId}`);
  if (!response.ok) {
    const err = (await response.json().catch(() => null)) as { detail?: string } | null;
    throw new Error(err?.detail ?? "Failed to get task");
  }
  return (await response.json()) as LocalTask;
}

export async function cancelTask(taskId: string): Promise<void> {
  const response = await authFetch(`/api/tasks/${taskId}/cancel`, { method: "POST" });
  if (!response.ok) {
    const err = (await response.json().catch(() => null)) as { detail?: string } | null;
    throw new Error(err?.detail ?? "Failed to cancel task");
  }
}

export async function deleteTask(taskId: string): Promise<void> {
  const response = await authFetch(`/api/tasks/${taskId}`, { method: "DELETE" });
  if (!response.ok) {
    const err = (await response.json().catch(() => null)) as { detail?: string } | null;
    throw new Error(err?.detail ?? "Failed to delete task");
  }
}

export async function getTaskLogs(taskId: string, tail = 100): Promise<TaskLogsResponse> {
  const response = await authFetch(`/api/tasks/${taskId}/logs?tail=${tail}`);
  if (!response.ok) {
    const err = (await response.json().catch(() => null)) as { detail?: string } | null;
    throw new Error(err?.detail ?? "Failed to get task logs");
  }
  return (await response.json()) as TaskLogsResponse;
}
